src/Npgsql/NpgsqlCommand.cs

```diff
@@ -740,31 +750,49 @@ async Task Unprepare(bool async, CancellationToken cancellationToken = default)
 
         #region Query analysis
 
-        internal void ProcessRawQuery(bool standardConformingStrings, bool deriveParameters)
+        internal void ProcessRawQuery(SqlQueryParser? parser, bool standardConformingStrings)
         {
             if (string.IsNullOrEmpty(CommandText))
                 throw new InvalidOperationException("CommandText property has not been initialized");
 
             NpgsqlStatement statement;
+
             switch (CommandType) {
             case CommandType.Text:
-                var parser = new SqlQueryParser();
-                parser.ParseRawQuery(CommandText, _parameters, _statements, standardConformingStrings, deriveParameters);
+                switch (Parameters.PlaceholderType)
+                {
+                case PlaceholderType.Positional:
+                    // In positional parameter mode, we don't need to parse/rewrite the CommandText or reorder the parameters - just use
+                    // them as is. If the SQL contains a semicolon (legacy batching) when positional parameters are in use, we just send
+                    // that and PostgreSQL will error (this behavior is by-design - use the new batching API).
+                    statement = TruncateStatementsToOne();
+                    statement.SQL = CommandText;
+                    statement.InputParameters = Parameters.InternalList;
+                    break;
+
+                case PlaceholderType.NoParameters:
+                    // Note that queries with no parameters are parsed just like queries with named parameters, since they may contain a
+                    // semicolon (legacy batching).
+                case PlaceholderType.Named:
+                    parser ??= new SqlQueryParser();
```

Comment:
So, I guess we do allocate a new parser for every multiplexing query without params?

Reply:
Yes, exactly. That's unfortunate - and this is one of the main reason that in the original attempt (#3834) I went in a more direction that tried to eliminate parsing for this case as well. But to summarize, in multiplexing:

* Before we enqueue to the multiplexing channel we have no connector - so no prepared SQL cache. So we can't use that to know whether the SQL contains semicolons or not (which was the general approach in #3834).
* After we enqueue to the multiplexing channel, we have a connector. But if we *do* have to parse/rewrite (e.g. named parameters), then we've moved parsing/rewriting to the critical bottleneck of the multiplexing loop. This is what caused the big perf regression in multi query with named parameters.

We can introduce an opt-in (app context switch, connection string param) that enforces raw SQL only - at that point the no-parameter case would also be optimized. We also have some future ideas for a global preparation cache (so not connector-based), in which case we can consider that too.

Reply:
I was talking about the parser allocations, not the parsing itself 🤣

Reply:
They both go together though :)

But yeah, we simply don't have a connector at this point, on which we would cache the SqlQueryParser instance. We could pool it some other way (or maybe cache on the command instance in case that's reused)... But assuming we introduce a raw-SQL opt-in, this all goes away anyway (we will never parse).